File: apps/dsa-web/src/store/quant.ts

import { create } from 'zustand';
import axios from 'axios';

const API_BASE = '/api/v1/quant';

interface Strategy {
  id: string;
  name: string;
  description: string;
  code: string;
  status: 'active' | 'inactive';
  created_at: string;
}

interface QuantStore {
  strategies: Strategy[];
  loading: boolean;
  generatedCode: string;
  
  fetchStrategies: () => Promise<void>;
  generateStrategy: (desc: string) => Promise<void>;
  saveStrategy: (name: string, desc: string, code: string) => Promise<void>;
  deleteStrategy: (id: string) => Promise<void>;
  toggleStrategy: (id: string, status: 'active' | 'inactive') => Promise<void>;
  clearGeneratedCode: () => void;
}

export const useQuantStore = create<QuantStore>((set, get) => ({
  strategies: [],
  loading: false,
  generatedCode: '',

  fetchStrategies: async () => {
    set({ loading: true });
    try {
      const res = await axios.get(`${API_BASE}/strategies`);
      set({ strategies: res.data });
    } catch (e) {
      console.error(e);
    } finally {
      set({ loading: false });
    }
  },

  generateStrategy: async (desc: string) => {
    set({ loading: true });
    try {
      const res = await axios.post(`${API_BASE}/strategies/generate`, { description: desc });
      set({ generatedCode: res.data.code });
    } catch (e) {
      console.error(e);
    } finally {
      set({ loading: false });
    }
  },

  saveStrategy: async (name, desc, code) => {
    try {
      await axios.post(`${API_BASE}/strategies`, { name, description: desc, code });
      get().fetchStrategies();
    } catch (e) {
      console.error(e);
    }
  },

  deleteStrategy: async (id) => {
    try {
      await axios.delete(`${API_BASE}/strategies/${id}`);
      get().fetchStrategies();
    } catch (e) {
      console.error(e);
    }
  },

  toggleStrategy: async (id, status) => {
    try {
      await axios.patch(`${API_BASE}/strategies/${id}/status`, { status });
      // Optimistic update
      set(state => ({
        strategies: state.strategies.map(s => 
          s.id === id ? { ...s, status } : s
        )
      }));
    } catch (e) {
      console.error(e);
      get().fetchStrategies(); // Revert on error
    }
  },
  
  clearGeneratedCode: () => set({ generatedCode: '' })
}));

File: api/v1/endpoints/quant.py
from fastapi import APIRouter, HTTPException, BackgroundTasks
from pydantic import BaseModel
from typing import List, Optional
import asyncio

# 引入我们之前的 quant 模块
from src.quant.strategy_manager import StrategyManager
from src.quant.runner import QuantRunner
from src.quant.monitor import RealTimeMonitor

# 全局单例 Monitor (为了简单，实际应放到 app state)
# ⚠️ 注意: Monitor 需要长时间运行，这里仅作 Demo 
# 实际生产中 Monitor 应是独立进程或后台任务
monitor_instance = RealTimeMonitor()

router = APIRouter()
strategy_mgr = StrategyManager()
runner = QuantRunner()

class StrategyCreateRequest(BaseModel):
    name: str
    description: str
    code: str

class StrategyGenerateRequest(BaseModel):
    description: str

class StrategyUpdateRequest(BaseModel):
    status: str # active / inactive

@router.get("/strategies", response_model=List[dict])
async def get_strategies():
    """获取所有策略"""
    return strategy_mgr.list_strategies()

@router.post("/strategies/generate")
async def generate_strategy(req: StrategyGenerateRequest):
    """根据描述生成策略代码"""
    code = await runner.build_strategy(req.description)
    return {"code": code}

@router.post("/strategies")
async def create_strategy(req: StrategyCreateRequest):
    """保存新策略"""
    strategy_id = strategy_mgr.add_strategy(req.name, req.description, req.code)
    return {"id": strategy_id, "message": "Strategy saved"}

@router.delete("/strategies/{strategy_id}")
async def delete_strategy(strategy_id: str):
    """删除策略"""
    success = strategy_mgr.delete_strategy(strategy_id)
    if not success:
        raise HTTPException(status_code=404, detail="Strategy not found")
    return {"message": "Deleted"}

@router.patch("/strategies/{strategy_id}/status")
async def update_strategy_status(strategy_id: str, req: StrategyUpdateRequest):
    """切换策略状态"""
    success = strategy_mgr.toggle_strategy(strategy_id, req.status)
    if not success:
        raise HTTPException(status_code=404, detail="Strategy not found")
    
    # 如果 monitor 正在运行，可能需要 reload
    if monitor_instance:
         # 简单的 reload context 触发更新
        monitor_instance.load_context()
        
    return {"message": f"Status updated to {req.status}"}

@router.post("/monitor/start")
async def start_monitor(background_tasks: BackgroundTasks):
    """启动监控 (Demo用)"""
    # 实际应该检查是否已运行
    # 这里只是演示 API 调用 monitor 逻辑
    return {"message": "Monitor start signal sent (Not fully implemented in API demo)"}
